### PR TITLE
[10.x] Add ability to set a custom class for the `AsCollection` and `AsEncryptedCollection` casts

### PR DESCRIPTION
**Pull request by osbre**
Example:
```php
protected $casts = [
    'elements' => AsCollection::class.':'.ElementCollection::class,
    // or
    'elements' => AsEncryptedCollection::class.':'.ElementCollection::class,
];
```

(In a similar way to `AsEnumCollection`)

```php
protected $casts = [
    'statuses' => AsEnumCollection::class.':'.ServerStatus::class,
];
```

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Casts/AsEncryptedCollection.php [hunk 1]
```diff
@@ -17,12 +17,21 @@ class AsEncryptedCollection implements Castable
...
     {
-        return new class implements CastsAttributes
+        return new class($arguments) implements CastsAttributes
         {
+            protected $arguments;
+
+            public function __construct(array $arguments)
+            {
+                $this->arguments = $arguments;
+            }
```
Comment (MelchiorKokernoot) [inline src/Illuminate/Database/Eloquent/Casts/AsEncryptedCollection.php:27]:
```suggestion

            public function __construct(protected array $arguments)
            {
            }
```

File: src/Illuminate/Database/Eloquent/Casts/AsCollection.php [hunk 2]
```diff
@@ -16,8 +16,15 @@ class AsCollection implements Castable
...
     {
-        return new class implements CastsAttributes
+        return new class($arguments) implements CastsAttributes
         {
+            protected $arguments;
+
+            public function __construct(array $arguments)
+            {
+                $this->arguments = $arguments;
+            }
```
[the remark this answers is not in the document]
Comment (osbre) [reply src/Illuminate/Database/Eloquent/Casts/AsCollection.php:26]: I did consider this syntax. However, I have not seen it being used in the framework before, which can lead to style inconsistencies.

File: src/Illuminate/Database/Eloquent/Casts/AsEncryptedCollection.php (hunk 1 above)
Comment (osbre) [reply src/Illuminate/Database/Eloquent/Casts/AsEncryptedCollection.php:27]: see comment above

File: src/Illuminate/Database/Eloquent/Casts/AsCollection.php [hunk 3]
```diff
@@ -26,7 +33,9 @@ public function get($model, $key, $value, $attributes)
 
                 $data = Json::decode($attributes[$key]);
 
-                return is_array($data) ? new Collection($data) : null;
+                $collectionClass = $this->arguments[0] ?? Collection::class;
```
[the remark this answers is not in the document]
Comment (osbre) [reply src/Illuminate/Database/Eloquent/Casts/AsCollection.php:36]: Attempted to do so in https://github.com/laravel/framework/pull/46619/commits/99c1465fd433963dfdd0552b806b2e53ea5664f5, which resulted in [tests failing](https://github.com/laravel/framework/actions/runs/4554449362/jobs/8032322295)

**Pull request merged**